contracts/NFTStorefrontV2.cdc

```diff
@@ -0,0 +1,705 @@
+                    assert(isCommissionRecipientValid, message: "Given recipient is not allowed")
+                }
+                let commissionPayment <- payment.withdraw(amount: self.details.commissionAmount)
+                let recipient = commissionRecipient.borrow() ?? panic("Unable to borrow the recipent capability")
+                recipient.deposit(from: <- commissionPayment)
+            }
+            // Fetch the token to return to the purchaser.
+            let nft <-self.nftProviderCapability.borrow()!.withdraw(withdrawID: self.details.nftID)
+            // Neither receivers nor providers are trustworthy, they must implement the correct
+            // interface but beyond complying with its pre/post conditions they are not gauranteed
+            // to implement the functionality behind the interface in any given way.
+            // Therefore we cannot trust the Collection resource behind the interface,
+            // and we must check the NFT resource it gives us to make sure that it is the correct one.
+            assert(nft.isInstance(self.details.nftType), message: "withdrawn NFT is not of specified type")
+            assert(nft.id == self.details.nftID, message: "withdrawn NFT does not have specified ID")
+
+            // Fetch the duplicate listing for the given NFT
+            // Access the StoreFrontManager resource reference to remove the duplicate listings if purchase would happen successfully.
+            let storeFrontPublicRef = self.owner!.getCapability<&NFTStorefrontV2.Storefront{NFTStorefrontV2.StorefrontPublic}>(NFTStorefrontV2.StorefrontPublicPath)
+                                        .borrow() ?? panic("Unable to borrow the storeFrontManager resource")
+            let duplicateListings = storeFrontPublicRef.getDuplicateListingIDs(nftType: self.details.nftType, nftID: self.details.nftID, listingID: self.uuid)
+
+            // Let's force removal of the listing in this storefront for the NFT that is being purchased. 
+            for listingID in duplicateListings {
+                storeFrontPublicRef.cleanup(listingResourceID: listingID)
+            }
+
+            // Rather than aborting the transaction if any receiver is absent when we try to pay it,
+            // we send the cut to the first valid receiver.
+            // The first receiver should therefore either be the seller, or an agreed recipient for
+            // any unpaid cuts.
+            var residualReceiver: &{FungibleToken.Receiver}? = nil
+            // Pay the comission 
+            // Pay each beneficiary their amount of the payment.
+
+            for cut in self.details.saleCuts {
+                if let receiver = cut.receiver.borrow() {
+                   let paymentCut <- payment.withdraw(amount: cut.amount)
+                    receiver.deposit(from: <-paymentCut)
+                    if (residualReceiver == nil) {
+                        residualReceiver = receiver
+                    }
+                }
```

Comment:
What about emitting an event in the else here so that it is possible to see what receivers that where invalid was supposed to get a cut?

Reply:
Yeah, that is a good idea but isn't this data can be generated off-chain by comparing the cuts receiver and the deposit events emitted during the transaction execution?

Reply:
Do we really want to rely on complicated off-chain dependent solutions where we can have an on-chain solution? This is not hard, and it will yield a way better more traceable solution.